rfcs/0000-open-arch.md

```diff
@@ -0,0 +1,225 @@
+> fairly open moving forward. As of now, this would be the deployment model for _our own_ *Metapolicies*
+
+To deploy our custom reconciler, we need to let the Kuadrant operator know about it, by adding it to the `Kuadrant` CR:
+
+```yaml
+apiVersion: kuadrant.io/v1beta2
+kind: Kuadrant
+metadata:
+  name: kuadrant-sample
+spec:
+  plugins:
+    userMetaPolicy:
+      deployment: embedded
+```
+
+The `UserMetaPolicy` will need to be packaged with a custom Docker image containing the plugin. It'll be automatically
+registered with the Kuadrant Controller from its own config file.
+
+> [!TODO]
+> Support out-of-image deployments of plugins, i.e. ones running as their own container. Which then also means we need
+> to deal with failures et al, so probably adding the necessity for more configuration options for these plugins
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Plugin architecture
+
+### gRPC to controller
+
+- [ ] Unix socket for "in-pod/embedded" plugins
+- [ ] Declarative way to load plugins
+- [ ] Provide a decorated [Controller Runtime](https://github.com/kubernetes-sigs/controller-runtime) Golang that
+  provides the changes triggered by the DAG
+
+### gRPC streams for eventing on the DAG
+
+- [ ] Subscriptions infered from CEL and the current CR being reconcilied
+
+### Typed DSL in CEL
+
+> [!IMPORTANT]
+> Provide both the `MetapolicyCR` & the `GwAPIContext` instances relative to a reconciliation
+> The `GwAPIContext` would not only be responsible for resolving the [CEL expressions][3] but also deriving the
+> necessary subscriptions needed to inform the controller when a reconciliation is needed because of changes to the
+> Gateway API objects queried
+
+- [ ] Support existing `context`s in the CEL interpreter's `Env`
+- [ ] Find a way to express the `context`s present/set for each CEL
+
+
+### Deployment
+
+> [!IMPORTANT]
+> This would initially be deployed as their own processes running alongside the kuadrant controller. They'd communicate
+> though unix sockets, over gRPC & gRPC streams (for the eventing).
+>
+> Later, we could support different deployment models, as all that'd be needed is being able to use gRPC essentially.
+
+- [ ] Known location for plugins to live in; one per directory
+- [ ] Unix socket opened within the same location?
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+> [!todo]
+> Why should we *not* do this?
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+> [!todo]
+> - Why is this design the best in the space of possible designs?
+> - What other designs have been considered and what is the rationale for not choosing them?
+> - What is the impact of not doing this?
+
+# Prior art
+[prior-art]: #prior-art
+
+> [!todo]
+> Discuss prior art, both the good and the bad, in relation to this proposal.
+> A few examples of what this can include are:
+>
+> - Does another project have a similar feature?
```

Comment:
Similar to ext_proc in envoy proxy? https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/ext_proc_filter